nixos/modules/services/misc/pufferpanel.nix

```diff
@@ -0,0 +1,621 @@
+{ config, pkgs, lib, ... }:
+let
+  cfg = config.services.pufferpanel;
+
+  ipwrap = address:
+    if lib.hasInfix ":" address
+    then "[" + address + "]"
+    else address;
+
+  joinHostPort = host: port:
+    "${ipwrap host}:${toString port}";
+
+  toStringOrNull = x:
+    if lib.isBool x then
+      lib.boolToString x
+    else if x != null then
+      toString x
+    else
+      null;
+
+  toJSONOrNull = x:
+    if x != null then
+      builtins.toJSON x
+    else
+      null;
+
+  mkOptionNullable = { type, default ? null, ... }@args:
+    lib.mkOption (args // {
+      type = lib.types.nullOr type;
+      inherit default;
+    });
+
+  options = lib.filterAttrs (_: v: v != null) ({
+    PUFFER_WEB_HOST = joinHostPort cfg.host cfg.port;
+    PUFFER_PANEL_ENABLE = lib.boolToString cfg.panel.enable;
+    PUFFER_DAEMON_ENABLE = lib.boolToString cfg.daemon.enable;
+  } // lib.optionalAttrs cfg.panel.enable {
+    PUFFER_PANEL_DATABASE_DIALECT = cfg.panel.database.dialect;
+    PUFFER_PANEL_DATABASE_LOG = toStringOrNull cfg.panel.database.log;
+    PUFFER_PANEL_WEB_FILES = cfg.panel.web.files;
+    PUFFER_PANEL_EMAIL_TEMPLATES = cfg.panel.email.templates;
+    PUFFER_PANEL_EMAIL_PROVIDER = cfg.panel.email.provider;
+    PUFFER_PANEL_EMAIL_FROM = cfg.panel.email.from;
+    PUFFER_PANEL_EMAIL_DOMAIN = cfg.panel.email.domain;
+    PUFFER_PANEL_EMAIL_HOST = cfg.panel.email.host;
+    PUFFER_PANEL_EMAIL_USERNAME = cfg.panel.email.username;
+    PUFFER_PANEL_SETTINGS_MASTERURL = cfg.panel.settings.masterUrl;
+    PUFFER_PANEL_SETTINGS_COMPANYNAME = cfg.panel.settings.companyName;
+    PUFFER_PANEL_SETTINGS_DEFAULTTHEME = cfg.panel.settings.defaultTheme;
+    PUFFER_PANEL_SETTINGS_THEMESETTINGS = toJSONOrNull cfg.panel.settings.themeSettings;
+    PUFFER_PANEL_REGISTRATIONENABLED = toStringOrNull cfg.panel.registrationEnabled;
+    PUFFER_TOKEN_PRIVATE = cfg.token.private;
+  } // lib.optionalAttrs cfg.daemon.enable {
+    PUFFER_DAEMON_CONSOLE_BUFFER = toStringOrNull cfg.daemon.console.buffer;
+    PUFFER_DAEMON_CONSOLE_FORWARD = toStringOrNull cfg.daemon.console.forward;
+    PUFFER_DAEMON_SFTP_HOST = joinHostPort cfg.daemon.sftp.host cfg.daemon.sftp.port;
+    PUFFER_DAEMON_SFTP_KEY = cfg.daemon.sftp.keyFile;
+    PUFFER_DAEMON_AUTH_URL = cfg.daemon.auth.url;
+    PUFFER_DAEMON_AUTH_CLIENTID = cfg.daemon.auth.clientId;
+    PUFFER_DAEMON_DATA_CRASHLIMIT = toStringOrNull cfg.daemon.data.crashLimit;
+    PUFFER_DAEMON_DATA_MAXWSDOWNLOADSIZE = toStringOrNull cfg.daemon.data.maxWSDownloadSize;
+    PUFFER_TOKEN_PUBLIC = cfg.token.public;
+  });
```

Comment:
Please see https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md.

Reply:
Thanks for pointing that out! You are right, per RFC, we can drop most of the options and let the user set environment variables, I’ll refactor this part.

Reply:
Should be good now.

Note that `token.public` is a common option (IIRC SFTP currently does not work if it is not set), but it’s deprecated and removed in PufferPanel’s `v3` branch. It’s possible to set public token URL using `PUFFER_TOKEN_PUBLIC` environment variable, so I didn’t add a NixOS option to avoid removing it once `v3` is released.